Add flag to declare token filters as updateable

cbuescher opened the pull request:
Currently token filter settings are treated as fixed once they are declared and
used in an analyzer. This is done to prevent changes in analyzers that are already
used actively to index documents, since changes to the analysis chain could
corrupt the index. However, it would be safe to allow updates to token
filters at search time ("search_analyzer"). This change introduces a new
property of token filters that allows to mark them as only being usable at search
or at index time. Any analyzer that uses these tokenfilters inherits that property 
and can be rejected if they are used in other contexts. This is a first step towards
making specific token filters (e.g. synonym filter) updateable.

Relates to #29051

Commit:
Add ability to declare filters/analyzers as updateable

Currently token filter settings are treated as fixed once they are declared and
used in an analyzer. This done to prevent changes in analyzers that are already
used actively to index documents, since changed to the analysis chain would
corrupt the index. However, it would be sage to allow updates to token
filters at search time ("search_analyzer"). This change introduces a new
property of token filters where they can now be declared to be "updatable". If
one tokenfilter in an analyzer is updateable in this way, it itself inherits
this property and we can reject these analyzers if they are tried to be used at
index time. This change demonstrates this with the "synonym" token filter as a
first example.

Relates to #29051

jimczi reviewed (commented):
I like the approach, I left some comments.

jimczi commented on server/src/main/java/org/elasticsearch/index/mapper/TypeParsers.java:
```diff
@@ -75,6 +75,10 @@ private static void parseAnalyzersAndTermVectors(FieldMapper.Builder builder, St
                 if (analyzer == null) {
                     throw new MapperParsingException("analyzer [" + propNode.toString() + "] not found for field [" + name + "]");
                 }
+                if (analyzer.isUpdateable()) {
```
We should check that the `default_analyzer` doesn't contain an updatable filters since it will be used by fields that don't declare an `analyzer`.

cbuescher replied:
Makes sense now, thanks. Will look into it.

jimczi commented on modules/analysis-common/src/test/resources/rest-api-spec/test/indices/10_updateable_synonyms.yml:
```diff
@@ -0,0 +1,95 @@
+                filter:
+                  syns_updateable:
+                    type: synonym
+                    synonyms: [ "quick,fast" ]
```
Sure I understand, just wanted to raise the question since we didn't discuss it.

Commit:
change from updatable flag to AnalysisMode enum

Commit:
Add detailed info about offending filters

Commit:
Add checks for default index analyzer

cbuescher commented:
@romseygeek @jimczi I added some commits changing from a boolean "updateable" flag to an enum (AnalysisMode), adding some more information to the error message if a search-time-only analyzer is detected at index time and also added a check for the default analyzer. Please take another look and let me know what you think are the next steps on this.

Commit:
Fix line length

jpountz commented on server/src/main/java/org/elasticsearch/index/analysis/AnalysisMode.java:
```diff
@@ -0,0 +1,74 @@
+    /**
+     * AnalysisMode representing analysis components that can be used only at index time
+     */
+    INDEX_TIME("index time") {
```
we don't seem to be using it anywhere? what analysis components would use this mode?

romseygeek replied:
Something like common grams has specific index- and query-time filters that this would be useful for.  We could also in future allow index-time synonym filters that only allow you to have single-term synonyms.

jpountz commented on server/src/main/java/org/elasticsearch/index/analysis/AnalysisMode.java:
```diff
@@ -0,0 +1,74 @@
+        return this.readableName;
+    }
+
+    public abstract AnalysisMode merge(AnalysisMode other);
```
can you add javadocs?

cbuescher replied:
done

jpountz reviewed (commented):
I like the approach too.

Commit:
javadoc

Commit:
Fix yaml tests

Commit:
Add updatable also to synonym_graph

Commit:
Merge branch 'master' into add-tokenfilter-updatability-flag

cbuescher commented:
@jimczi @romseygeek I just updated this PR so it passes on current master, I think the approach is still okay, would you mind taking another look? 
One thing to keep in mind this is just the part that adds a way to distiguish between token filters that are okay to be updated and those that are not. I remember I had a separate PR/branch where I added the actual updating process through a separate API call. That isn't included here and I'll have to dig that up again.

jimczi commented on modules/analysis-common/src/main/java/org/elasticsearch/analysis/common/SynonymTokenFilterFactory.java:
```diff
@@ -50,6 +51,7 @@
     private final boolean lenient;
     protected final Settings settings;
     protected final Environment environment;
+    private final boolean updateable;
```
Since this is a noop for now I think it would be better to test the functionality in a custom token filter only used in test. We can work on synonyms update as a follow up but we should not add this option if the logic is not implemented ?

cbuescher replied:
Good point, will see how that would look like and what that means for testing.

jimczi reviewed (changes requested):
I like the change and I think it's a good step forward. I left one comment regarding the usage of the new flag and how we should test it. One thing I wonder is how we could ensure that updates are not applied in the middle of a request but we can discuss this in the follow up pr that will handle updates in the synonym filter.

Commit:
Remove updateability flage from synonym token filters for now

Commit:
Merge branch 'master' of github.com:elastic/elasticsearch into add-tokenfilter-updatability-flag

Commit:
adding tests and moving some code

Commit:
iter

Commit:
adding tests

cbuescher commented:
@jimczi I added tests to we currently only use the new flag functionality in test token filters and removed it from the synonym token fitlers together with the rest. I think its ready for another look.

jimczi commented on server/src/test/java/org/elasticsearch/index/analysis/NamedAnalyzerTests.java:
```diff
@@ -0,0 +1,79 @@
+        }
+    }
+
+    private Analyzer createAnalyzerWithMode(String name, AnalysisMode mode) {
```
It's great that we can have a unit test for this but I wonder if an integ test with a analyzer plugin should also be tested ? This way you could test the high level behavior when you mix a custom token filter with specific analysis mode and other token filters that use the ALL mode ?

cbuescher replied:
Sorry, I don't really see the benefit of an integration test at this point. What would an analyzer plugin add that is not tested here?
As soon as we use different modes, e.g. like I was already doing previously in this PR with the SynonymTokenFilter, it will be tested in integration and rest tests, but I don't see any benefit adding it with other mocked filters here.

jimczi replied:
> What would an analyzer plugin add that is not tested here?

I think it would provide a good example of how to use and a way to check if the mode is correctly propagated when you mix multiple analyzer. I agree that most of this is already tested in the mock test so I don't have strong feeling about it either.

jimczi reviewed (commented):
Thanks for adding the tests @cbuescher , I left another comment to add an integration test for this.

Commit:
iter

jimczi commented on server/src/main/java/org/elasticsearch/index/mapper/TypeParsers.java:
```diff
@@ -73,20 +74,23 @@ private static void parseAnalyzersAndTermVectors(FieldMapper.Builder builder, St
                 if (analyzer == null) {
                     throw new MapperParsingException("analyzer [" + propNode.toString() + "] not found for field [" + name + "]");
                 }
+                analyzer.checkAllowedInMode(AnalysisMode.ALL);
```
Should we check the `AnalysisMode.INDEX_TIME` mode if `search_analyzer` is also set ?

cbuescher replied:
Just to clarify: you mean we should allow an "analyzer" to be restricted to "index time" only if the search analyzer is set?

jimczi replied:
I think so yes, if the analyzer is set to INDEX_TIME this means that it should not be used for queries so we should check that. My understanding is that the ALL mode check will fail if the analyzer is set to INDEX_TIME even though a search analyzer is defined, this looks like a bug to me.

cbuescher replied:
Yes, that makes sense. Good point, will add that.

Commit:
allow analyzer restricted to index time if search analyzer also set

Commit:
add test

cbuescher commented:
@jimczi I added the special case of "analyzer" an no "search_analyzer" set, this should throw an error if the "analyzer" is index-time only.

jimczi reviewed (approved):
> I added the special case of "analyzer" an no "search_analyzer" set, this should throw an error if the "analyzer" is index-time only.

Thanks, LGTM

Commit:
Fix line length

Commit:
Merge branch 'master' into add-tokenfilter-updatability-flag
